core/lib/contract_verifier/src/etherscan/mod.rs

```diff
@@ -0,0 +1,459 @@
+                );
+                return Ok(None);
+            }
+            Err(e) => return Err(e.into()),
+        };
+        let mut connection = self
+            .connection_pool
+            .connection_tagged("etherscan_verifier")
+            .await?;
+
+        connection
+            .etherscan_verification_dal()
+            .save_etherscan_verification_id(verification_request.id, &request_id)
+            .await?;
+
+        Ok(Some(request_id))
+    }
+
+    async fn await_verification(
+        &mut self,
+        verification_request: &VerificationRequest,
+        verification_id: String,
+    ) -> Result<(), VerifierError> {
+        let mut get_status_attempts = 0;
+        loop {
+            get_status_attempts += 1;
+            tracing::info!(
+                "Fetching contract verification status, address = {:#?}, attempt = {}",
+                verification_request.req.contract_address,
+                get_status_attempts
+            );
+            let result = self.client.get_verification_status(&verification_id).await;
+            match result {
+                Ok(_) => break,
+                Err(EtherscanError::VerificationPending) => {
+                    if tokio::time::timeout(
+                        POLL_VERIFICATION_RESULT_INTERVAL,
+                        self.stop_receiver.changed(),
+                    )
+                    .await
+                    .is_ok()
+                    {
+                        return Err(VerifierError::Canceled);
+                    }
+                    if get_status_attempts > MAX_POLL_VERIFICATION_RESULT_ATTEMPTS {
+                        tracing::warn!(
+                            "Get verification status timed out for contract with address {:#?}",
+                            verification_request.req.contract_address
+                        );
+                        return Err(ProcessingError::VerificationStatusPollingTimeout.into());
+                    }
+                    continue;
+                }
+                Err(e) => return Err(e.into()),
+            }
+        }
+
+        Ok(())
+    }
+
+    async fn verify(
+        &mut self,
+        verification_request: VerificationRequest,
+        etherscan_verification_id: Option<String>,
+    ) -> Result<(), VerifierError> {
+        let req = verification_request.req.clone();
+        let is_verified = self
+            .client
+            .is_contract_verified(req.contract_address)
+            .await?;
+
+        if is_verified {
+            tracing::info!(
+                "Contract with address {:#?} is already verified",
+                req.contract_address
+            );
+            return Ok(());
+        }
+        let verification_id = match etherscan_verification_id {
+            Some(id) => Some(id),
+            None => {
+                let request_id = self
+                    .send_verification_request(&verification_request)
+                    .await?;
```

Comment:
Can we do

```rust
let Some(request_id) = self.send_verification_request(&verification_request).await? else {
    // Already verified.
    return Ok(());
}
```

here, so that below we don't have to deal with `Option`?